runtime/callback.c

```diff
@@ -62,9 +62,9 @@ static inline void restore_stack_parent(caml_domain_state* domain_state, value c
 #include "caml/fix_code.h"
 #include "caml/fiber.h"
 
-static opcode_t callback_code[] = { ACC, 0, APPLY, 0, POP, 1, STOP };
+static __thread opcode_t callback_code[] = { ACC, 0, APPLY, 0, POP, 1, STOP };
 
-static int callback_code_inited = 0;
+static __thread int callback_code_inited = 0;
```

Comment:
Do these need to be thread-local variables? Do we need to initialise callback_code for each domain? 

Looking at the `caml_register_code_fragment` [1], it appears not to be thread-safe; I don't think `caml_skiplist_insert` is. And making these variables thread-local does not solve the problem. 

It may be better to add a lock around [1] and remove `__thread` annotation. 

[1] https://github.com/ocaml-multicore/ocaml-multicore/blob/4.12%2Bdomains%2Beffects/runtime/codefrag.c#L56-L60

Reply:
You're right `codefrag.c` isn't thread-safe; we need to fix that, we could have multiple domains using dynlink for example. 

For the specific usage in callback, there are a couple of things we need to achieve:
- `callback_code` needs to be registered as a codefrag as it is now appearing in the stack (see `caml_callbackN_exn`) for the return
- we need to make a call on how `callback_code` is registered with codefrag

I went with the following strategy for how `callback_code` is registered:
 - make each codefrag area a `__thread` with intitialization lazy and at most once using `callback_code_inited`
 
The alternate solution I could see was:
 - copy out of the global area and use a stack area, but then we take the overhead of registering and deregistering the codefrag on every callback

I think we want to avoid hitting codefrag `register`/`deregister` on every callback which is why I went with the `__thread`. Does that make sense? Or are there other options better than those?